apps/explorer/lib/explorer/chain/cache/gas_price_oracle.ex

```diff
@@ -17,148 +17,251 @@ defmodule Explorer.Chain.Cache.GasPriceOracle do
     Wei
   }
 
-  alias Explorer.Repo
+  alias Explorer.Counters.AverageBlockTime
+  alias Explorer.{Market, Repo}
+  alias Timex.Duration
 
   use Explorer.Chain.MapCache,
     name: :gas_price,
     key: :gas_prices,
+    key: :gas_prices_acc,
+    key: :updated_at,
     key: :old_gas_prices,
     key: :async_task,
-    global_ttl: Application.get_env(:explorer, __MODULE__)[:global_ttl],
+    global_ttl: global_ttl(),
     ttl_check_interval: :timer.seconds(1),
     callback: &async_task_on_deletion(&1)
 
   @doc """
   Get `safelow`, `average` and `fast` percentile of transactions gas prices among the last `num_of_blocks` blocks
   """
   @spec get_average_gas_price(pos_integer(), pos_integer(), pos_integer(), pos_integer()) ::
-          {:error, any} | {:ok, %{String.t() => nil | float, String.t() => nil | float, String.t() => nil | float}}
+          {{:error, any} | {:ok, %{slow: gas_price, average: gas_price, fast: gas_price}},
+           [
+             %{
+               block_number: non_neg_integer(),
+               slow_gas_price: nil | Decimal.t(),
+               fast_gas_price: nil | Decimal.t(),
+               average_gas_price: nil | Decimal.t(),
+               slow_priority_fee_per_gas: nil | Decimal.t(),
+               average_priority_fee_per_gas: nil | Decimal.t(),
+               fast_priority_fee_per_gas: nil | Decimal.t(),
+               slow_time: nil | Decimal.t(),
+               average_time: nil | Decimal.t(),
+               fast_time: nil | Decimal.t()
+             }
+           ]}
+        when gas_price: nil | %{price: float(), time: float(), fiat_price: Decimal.t()}
   def get_average_gas_price(num_of_blocks, safelow_percentile, average_percentile, fast_percentile) do
     safelow_percentile_fraction = safelow_percentile / 100
     average_percentile_fraction = average_percentile / 100
     fast_percentile_fraction = fast_percentile / 100
 
+    acc = get_gas_prices_acc()
+
+    from_block =
+      case acc do
+        [%{block_number: from_block} | _] -> from_block
+        _ -> -1
+      end
+
+    average_block_time =
+      case AverageBlockTime.average_block_time() do
+        {:error, _} -> nil
+        average_block_time -> average_block_time |> Duration.to_milliseconds()
+      end
+
     fee_query =
       from(
         block in Block,
         left_join: transaction in assoc(block, :transactions),
         where: block.consensus == true,
         where: transaction.status == ^1,
         where: transaction.gas_price > ^0,
+        where: block.number > ^from_block,
```

Comment:
can we use `transaction.block_number` instead taking in mind upcoming `transactions` table denormalization?

Reply:
Btw, what is the rationale for this extra condition?

Reply:
As this update assumes more frequent updates, we do not want to query the database for the whole blocks range each time, so we take only those that we do not have in accumulator

Reply:
Ok, how about to change to for seamless integration with upcoming transactions table denormalization:

```suggestion
        where: transaction.block_number > ^from_block,
```

Reply:
Done [here](https://github.com/blockscout/blockscout/pull/9044/commits/79941aea62431fdb75897dd6ea8aac578a740f83)

I guess we need to add here `if` with denormalization helper condition to eliminate blocks join when the denormalization will be in master